syntax = "proto3";

package raftNode;

service RaftNodeService {
  rpc RequestVote (RequestVoteRequest) returns (RequestVoteResponse) {}
  rpc AppendEntries (AppendEntriesRequest) returns (AppendEntriesResponse) {}
  rpc ServeClient (ServeClientArgs) returns (ServeClientResponse) {}
}

message RequestVoteRequest {
  int32 term = 1;
  int32 candidateId = 2;
  int32 lastLogIndex = 3;
  int32 lastLogTerm = 4;
}

message RequestVoteResponse {
  int32 term = 1;
  bool voteGranted = 2;
}

message AppendEntriesRequest {
  int32 term = 1;
  int32 leaderId = 2;
  int32 prevLogIndex = 3;
  int32 prevLogTerm = 4;
  repeated LogEntry entries = 5;
  int32 leaderCommit = 6;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
}

message LogEntry {
  int32 term = 1;
  string command = 2;
}

message ServeClientArgs {
  string request = 1;
}

message ServeClientSuccessReply {
  string data = 1;
  string leaderId = 2;
  bool success = 3;
}

message ServeClientFailureReply {
  string data = 1;
  string leaderId = 2;
  bool success = 3;
}

message ServeClientResponse {
  oneof reply {
    ServeClientSuccessReply successReply = 1;
    ServeClientFailureReply failureReply = 2;
  }
}
